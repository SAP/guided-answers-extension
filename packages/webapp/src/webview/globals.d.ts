interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
