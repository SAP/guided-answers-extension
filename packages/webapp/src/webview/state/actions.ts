export {
    updateGuidedAnswerTrees,
    updateActiveNode,
    selectNode,
    goToPreviousPage,
    goToAllAnswers,
    restartAnswer,
    executeCommand,
    setActiveTree,
    searchTree,
    setQueryValue,
    webviewReady,
    getBetaFeatures,
    sendFeedbackOutcome,
    sendFeedbackComment,
    guideFeedback,
    setProductFilters,
    setComponentFilters,
    resetFilters,
    setPageSize,
    feedbackStatus,
    feedbackResponse
} from '@sap/guided-answers-extension-types';
