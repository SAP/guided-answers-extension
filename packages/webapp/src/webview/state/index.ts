export { actions, store } from './store';
export { getInitialState } from './reducers';
