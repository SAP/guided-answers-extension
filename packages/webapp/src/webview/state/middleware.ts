import i18next from 'i18next';
import type { Middleware, MiddlewareAPI, Dispatch, Action } from 'redux';
import type { GuidedAnswerActions } from '@sap/guided-answers-extension-types';
import type { AppState } from '../types';

declare let window: Window;
declare let acquireVsCodeApi: () => typeof window['vscode'];

/**
 * Communication between IDE extension and web view is realized through the communication middleware
 * window event listener for incoming messages in to the webview
 * post message for outgoing messages to the host.
 *
 * @param store - redux store
 * @returns - redux middleware
 */
export const communicationMiddleware: Middleware<
    Dispatch<GuidedAnswerActions>,
    AppState,
    Dispatch<GuidedAnswerActions>
> = (store: MiddlewareAPI) => {
    // Add event handler, this will dispatch incomming state updates
    window.addEventListener('message', (event: MessageEvent) => {
        if (event.origin === window.origin) {
            console.log(i18next.t('MESSAGE_RECEIVED'), event);
            if (event.data && typeof event.data.type === 'string') {
                store.dispatch(event.data);
            }
        } else {
            console.error(i18next.t('UNKNOWN_ORIGIN') + event.origin);
        }
    });
    try {
        window.vscode = acquireVsCodeApi();
    } catch (e) {
        console.error(i18next.t('ERROR_NOT_RUNNING_IN_WEB_VIEW'));
    }
    return (next: Dispatch<GuidedAnswerActions>) =>
        (action): Action => {
            action = next(action);
            if (action && typeof action.type === 'string' && !action.type.startsWith('[view]')) {
                window.vscode.postMessage(action);
                console.log(i18next.t('REACT_ACTION_POSTED'), action);
            }
            return action;
        };
};
