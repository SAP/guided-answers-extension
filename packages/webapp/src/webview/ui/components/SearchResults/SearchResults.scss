.guided-answer__search-results {
    &__node-list {
        padding-bottom: 10px;

        >ul {
            padding: 0;
        }
    }

    &__separator {
        margin: 0 15px 5px 15px;
        border: none;
        border-top: 1px solid var(--vscode-editorWidget-border);
    }

    &__view-more {
        margin: 10px 0 0 30px;
        cursor: pointer;

        &:hover {
            >span {
                text-decoration-line: none;
            }
        }

        >i {
            height: 16px;
            margin-right: 5px;
            vertical-align: middle;
        }

        >span {
            color: var(--vscode-foreground);
            font-size: 11px;
            font-weight: 400;
            line-height: 16px;
            text-decoration-line: underline;
            vertical-align: middle;
        }
    }

    &__node {
        list-style-type: none;

        >button {
            display: flex;
            width: 100%;
            padding: 5px 15px 5px 15px;

            &:hover {
                background-color: var(--vscode-list-hoverBackground);
                outline: 1px solid var(--vscode-focusBorder, #007fd4);
                outline-offset: -1px;
            }

            &:focus-visible {
                outline: 1px solid var(--vscode-focusBorder, #007fd4);
                outline-offset: -2px;
            }
        }

        &__indent {
            width: 10px;
            height: 5px;
            margin: 3px 5px 0 0;
            border-bottom: 1px solid var(--vscode-editorWidget-border);
            border-left: 1px solid var(--vscode-editorWidget-border);
        }

        &__title {
            display: flex;
            align-items: center;
            gap: 5px;
            margin: 0 0 5px 0;
            padding: 0;
            font-size: 11px;
            line-height: 16px;
            color: var(--vscode-settings-headerForeground);
            text-align: left;

            >i {
                height: 16px;
            }
        }

        &__detail {
            margin: 0;
            padding: 0;
            font-size: 11px;
            line-height: 16px;
            color: var(--vscode-foreground);
            text-align: left;
        }
    }
}
