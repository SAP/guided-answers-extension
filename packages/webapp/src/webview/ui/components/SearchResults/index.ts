export * from './SearchResults';
export * from './SearchResultsTree';
