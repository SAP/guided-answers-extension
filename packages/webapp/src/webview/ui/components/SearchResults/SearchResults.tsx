import type { ReactElement } from 'react';
import React from 'react';
import { useSelector } from 'react-redux';
import InfiniteScroll from 'react-infinite-scroll-component';
import i18next from 'i18next';
import { FocusZone, FocusZoneDirection } from '@fluentui/react-focus';
import { SpinnerSize } from '@fluentui/react';
import { UILoader } from '@sap-ux/ui-components';
import { actions } from '../../../state';
import type { AppState } from '../../../types';
import { ErrorScreen } from '../ErrorScreen';
import { SearchResultsTree } from './SearchResultsTree';
import './SearchResults.scss';

/**
 * Search results screen for Guided Answers Extension app.
 *
 * @returns - react elements for the search results
 */
export function SearchResults(): ReactElement {
    const appState = useSelector<AppState, AppState>((state) => state);

    const fetchData = () => {
        if (appState.guidedAnswerTreeSearchResult.resultSize > appState.pageSize) {
            actions.searchTree({
                query: appState.query,
                filters: {
                    product: appState.selectedProductFilters,
                    component: appState.selectedComponentFilters
                },
                paging: {
                    responseSize: appState.pageSize,
                    offset: appState.guidedAnswerTreeSearchResult.trees.length
                }
            });
        }
    };

    return appState.guidedAnswerTreeSearchResult.resultSize === 0 ? (
        <ErrorScreen title={i18next.t('NO_ANSWERS_FOUND')} subtitle={i18next.t('PLEASE_MODIFY_SEARCH')} />
    ) : (
        <FocusZone direction={FocusZoneDirection.bidirectional} isCircularNavigation={true}>
            <ul className="striped-list" role="listbox">
                <InfiniteScroll
                    dataLength={appState.guidedAnswerTreeSearchResult.trees.length} //This is important field to render the next data
                    next={fetchData}
                    loader={<UILoader id="loading-indicator" size={SpinnerSize.large} />}
                    hasMore={
                        appState.guidedAnswerTreeSearchResult.trees.length <
                        appState.guidedAnswerTreeSearchResult.resultSize
                    }>
                    {appState.guidedAnswerTreeSearchResult.trees.map((tree) => (
                        <SearchResultsTree key={`tree-item-${tree.TITLE}`} tree={tree} />
                    ))}
                </InfiniteScroll>
            </ul>
        </FocusZone>
    );
}
