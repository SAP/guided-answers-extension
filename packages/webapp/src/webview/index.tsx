import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
