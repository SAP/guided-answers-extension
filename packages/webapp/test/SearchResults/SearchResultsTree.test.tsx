import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { fireEvent } from '@testing-library/dom';
import { useSelector } from 'react-redux';
import type { GuidedAnswerTreeSearchHit } from '@sap/guided-answers-extension-types';
import { actions } from '../../src/webview/state';
import { SearchResultsTree } from '../../src/webview/ui/components/SearchResults';

jest.mock('../../src/webview/state', () => ({
    actions: {
        setActiveTree: jest.fn(),
        selectNode: jest.fn(),
        expandSearchNodesForTree: jest.fn()
    }
}));

jest.mock('react-redux', () => ({
    useSelector: jest.fn()
}));

describe('<SearchResultsTree />', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                bookmarks: { '1-12': {} },
                searchResultVisibleNodeCount: {
                    1: 1
                }
            })
        );
        const tree = {
            TITLE: 'Tree title',
            DESCRIPTION: 'Tree description',
            PRODUCT: 'product',
            COMPONENT: 'component',
            TREE_ID: 1,
            FIRST_NODE_ID: 11,
            ACTIONS: [
                {
                    TREE_ID: 1,
                    NODE_ID: 12,
                    TITLE: 'Node title',
                    DETAIL: 'Node detail'
                }
            ]
        } as GuidedAnswerTreeSearchHit;

        const { container } = render(<SearchResultsTree tree={tree} />);
        expect(container).toMatchSnapshot();
    });

    it('opens tree', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                bookmarks: {},
                searchResultVisibleNodeCount: {
                    1: 1
                }
            })
        );
        const tree = {
            TITLE: 'Tree title',
            DESCRIPTION: 'Tree description',
            PRODUCT: 'product',
            COMPONENT: 'component',
            TREE_ID: 1,
            FIRST_NODE_ID: 11,
            ACTIONS: [
                {
                    TREE_ID: 1,
                    NODE_ID: 12,
                    TITLE: 'Node title',
                    DETAIL: 'Node detail'
                }
            ]
        } as GuidedAnswerTreeSearchHit;

        const { container } = render(<SearchResultsTree tree={tree} />);
        expect(container).toMatchSnapshot();
        fireEvent.click(screen.getByTestId('search-result-tree-button'));
        expect(actions.setActiveTree).toBeCalledWith(tree);
        expect(actions.selectNode).toBeCalledWith(11);
    });

    it('expands tree nodes', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                bookmarks: {},
                searchResultVisibleNodeCount: {
                    1: 1
                }
            })
        );
        const tree = {
            TITLE: 'Tree title',
            DESCRIPTION: 'Tree description',
            PRODUCT: 'product',
            COMPONENT: 'component',
            TREE_ID: 1,
            FIRST_NODE_ID: 11,
            ACTIONS: [
                {
                    TREE_ID: 1,
                    NODE_ID: 12,
                    TITLE: 'Node title 1',
                    DETAIL: 'Node detail 1'
                },
                {
                    TREE_ID: 1,
                    NODE_ID: 13,
                    TITLE: 'Node title 2',
                    DETAIL: 'Node detail 2'
                }
            ]
        } as GuidedAnswerTreeSearchHit;

        const { container } = render(<SearchResultsTree tree={tree} />);
        expect(container).toMatchSnapshot();
        fireEvent.click(screen.getByTestId('search-result-tree-expand-button'));
        expect(actions.expandSearchNodesForTree).toBeCalledWith(1);
    });

    it('opens tree node', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                bookmarks: {},
                searchResultVisibleNodeCount: {
                    1: 1
                }
            })
        );
        const tree = {
            TITLE: 'Tree title',
            DESCRIPTION: 'Tree description',
            PRODUCT: 'product',
            COMPONENT: 'component',
            TREE_ID: 1,
            FIRST_NODE_ID: 11,
            ACTIONS: [
                {
                    TREE_ID: 1,
                    NODE_ID: 12,
                    TITLE: 'Node title',
                    DETAIL: 'Node detail'
                }
            ]
        } as GuidedAnswerTreeSearchHit;

        const { container } = render(<SearchResultsTree tree={tree} />);
        expect(container).toMatchSnapshot();
        fireEvent.click(screen.getByTestId('search-result-node-button'));
        expect(actions.setActiveTree).toBeCalledWith(tree);
        expect(actions.selectNode).toBeCalledWith(12);
    });
});
