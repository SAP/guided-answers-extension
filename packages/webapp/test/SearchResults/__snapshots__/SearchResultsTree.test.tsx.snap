// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultsTree /> expands tree nodes 1`] = `
<div>
  <li
    class="tree-item"
    role="option"
  >
    <button
      class="guided-answer__tree"
      id="search-result-tree-button"
    >
      <div
        class="guided-answer__tree__ul"
      >
        <h3
          class="guided-answer__tree__title"
        >
          Tree title
           
          
        </h3>
        <div
          class="bottom-section"
          id="bottom-section"
        >
          <span
            class="guided-answer__tree__desc"
          >
            Tree description
          </span>
          <div
            class="component-and-product-container"
            style="margin-top: 10px;"
          >
            <div
              class="guided-answer__tree__product"
              id="product-container"
            >
              <span
                class="bottom-title"
              >
                Product: 
              </span>
              product
            </div>
            <div
              class="guided-answer__tree__component"
              id="component-container"
            >
              <span
                class="bottom-title"
              >
                Component: 
              </span>
              component
            </div>
          </div>
        </div>
      </div>
    </button>
    <div
      class="guided-answer__search-results__node-list"
    >
      <hr
        class="guided-answer__search-results__separator"
      />
      <ul
        role="listbox"
      >
        <li
          class="guided-answer__search-results__node"
          role="option"
        >
          <button
            id="search-result-node-button"
          >
            <div
              class="guided-answer__search-results__node__indent"
            />
            <div>
              <h3
                class="guided-answer__search-results__node__title"
              >
                <span>
                  Node title 1
                </span>
                
              </h3>
              <p
                class="guided-answer__search-results__node__detail"
              >
                Node detail 1
              </p>
            </div>
          </button>
        </li>
      </ul>
      <a
        class="guided-answer__search-results__view-more"
        id="search-result-tree-expand-button"
      >
        <i
          aria-hidden="true"
          class="ts-icon root-105"
          data-icon-name="ContractNodesSmall"
        />
        <span />
      </a>
    </div>
  </li>
</div>
`;

exports[`<SearchResultsTree /> opens tree 1`] = `
<div>
  <li
    class="tree-item"
    role="option"
  >
    <button
      class="guided-answer__tree"
      id="search-result-tree-button"
    >
      <div
        class="guided-answer__tree__ul"
      >
        <h3
          class="guided-answer__tree__title"
        >
          Tree title
           
          
        </h3>
        <div
          class="bottom-section"
          id="bottom-section"
        >
          <span
            class="guided-answer__tree__desc"
          >
            Tree description
          </span>
          <div
            class="component-and-product-container"
            style="margin-top: 10px;"
          >
            <div
              class="guided-answer__tree__product"
              id="product-container"
            >
              <span
                class="bottom-title"
              >
                Product: 
              </span>
              product
            </div>
            <div
              class="guided-answer__tree__component"
              id="component-container"
            >
              <span
                class="bottom-title"
              >
                Component: 
              </span>
              component
            </div>
          </div>
        </div>
      </div>
    </button>
    <div
      class="guided-answer__search-results__node-list"
    >
      <hr
        class="guided-answer__search-results__separator"
      />
      <ul
        role="listbox"
      >
        <li
          class="guided-answer__search-results__node"
          role="option"
        >
          <button
            id="search-result-node-button"
          >
            <div
              class="guided-answer__search-results__node__indent"
            />
            <div>
              <h3
                class="guided-answer__search-results__node__title"
              >
                <span>
                  Node title
                </span>
                
              </h3>
              <p
                class="guided-answer__search-results__node__detail"
              >
                Node detail
              </p>
            </div>
          </button>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`<SearchResultsTree /> opens tree node 1`] = `
<div>
  <li
    class="tree-item"
    role="option"
  >
    <button
      class="guided-answer__tree"
      id="search-result-tree-button"
    >
      <div
        class="guided-answer__tree__ul"
      >
        <h3
          class="guided-answer__tree__title"
        >
          Tree title
           
          
        </h3>
        <div
          class="bottom-section"
          id="bottom-section"
        >
          <span
            class="guided-answer__tree__desc"
          >
            Tree description
          </span>
          <div
            class="component-and-product-container"
            style="margin-top: 10px;"
          >
            <div
              class="guided-answer__tree__product"
              id="product-container"
            >
              <span
                class="bottom-title"
              >
                Product: 
              </span>
              product
            </div>
            <div
              class="guided-answer__tree__component"
              id="component-container"
            >
              <span
                class="bottom-title"
              >
                Component: 
              </span>
              component
            </div>
          </div>
        </div>
      </div>
    </button>
    <div
      class="guided-answer__search-results__node-list"
    >
      <hr
        class="guided-answer__search-results__separator"
      />
      <ul
        role="listbox"
      >
        <li
          class="guided-answer__search-results__node"
          role="option"
        >
          <button
            id="search-result-node-button"
          >
            <div
              class="guided-answer__search-results__node__indent"
            />
            <div>
              <h3
                class="guided-answer__search-results__node__title"
              >
                <span>
                  Node title
                </span>
                
              </h3>
              <p
                class="guided-answer__search-results__node__detail"
              >
                Node detail
              </p>
            </div>
          </button>
        </li>
      </ul>
    </div>
  </li>
</div>
`;

exports[`<SearchResultsTree /> renders without crashing 1`] = `
<div>
  <li
    class="tree-item"
    role="option"
  >
    <button
      class="guided-answer__tree"
      id="search-result-tree-button"
    >
      <div
        class="guided-answer__tree__ul"
      >
        <h3
          class="guided-answer__tree__title"
        >
          Tree title
           
          <i
            aria-hidden="true"
            class="ts-icon root-105"
            data-icon-name="StarActive"
          />
        </h3>
        <div
          class="bottom-section"
          id="bottom-section"
        >
          <span
            class="guided-answer__tree__desc"
          >
            Tree description
          </span>
          <div
            class="component-and-product-container"
            style="margin-top: 10px;"
          >
            <div
              class="guided-answer__tree__product"
              id="product-container"
            >
              <span
                class="bottom-title"
              >
                Product: 
              </span>
              product
            </div>
            <div
              class="guided-answer__tree__component"
              id="component-container"
            >
              <span
                class="bottom-title"
              >
                Component: 
              </span>
              component
            </div>
          </div>
        </div>
      </div>
    </button>
    <div
      class="guided-answer__search-results__node-list"
    >
      <hr
        class="guided-answer__search-results__separator"
      />
      <ul
        role="listbox"
      >
        <li
          class="guided-answer__search-results__node"
          role="option"
        >
          <button
            id="search-result-node-button"
          >
            <div
              class="guided-answer__search-results__node__indent"
            />
            <div>
              <h3
                class="guided-answer__search-results__node__title"
              >
                <span>
                  Node title
                </span>
                <i
                  aria-hidden="true"
                  class="ts-icon root-105"
                  data-icon-name="StarActive"
                />
              </h3>
              <p
                class="guided-answer__search-results__node__detail"
              >
                Node detail
              </p>
            </div>
          </button>
        </li>
      </ul>
    </div>
  </li>
</div>
`;
